torch/nativert/graph/GraphSignature.cpp

```diff
@@ -0,0 +1,505 @@
+          torch::_export::Argument::Tag::AS_CUSTOM_OBJ) {
+        argName =
+            inputSpec.get_user_input().get_arg().get_as_custom_obj().get_name();
+      } else {
+        TORCH_CHECK(false, "Unsupported USER_INPUT argument type.");
+      }
+      break;
+    case torch::_export::InputSpec::Tag::CONSTANT_INPUT:
+      argName = inputSpec.get_constant_input().get_name();
+      tagName = "CONSTANT_INPUT";
+      break;
+    case torch::_export::InputSpec::Tag::TOKEN:
+      TORCH_CHECK(false, "Token inputs not implemented yet.");
+    default:
+      TORCH_CHECK(false, "Unknown InputSpec tag encountered.");
+  }
+  return std::make_pair(argName, tagName);
+}
+
+void checkInputOrders(
+    const std::vector<torch::_export::InputSpec>& inputSpecs) {
+  // Map each tag to its index in the expected order
+  std::unordered_map<torch::_export::InputSpec::Tag, size_t> tagOrderMap = {
+      {torch::_export::InputSpec::Tag::TOKEN, 0},
+      {torch::_export::InputSpec::Tag::PARAMETER, 1},
+      {torch::_export::InputSpec::Tag::BUFFER, 2},
+      {torch::_export::InputSpec::Tag::TENSOR_CONSTANT, 3},
+      {torch::_export::InputSpec::Tag::CUSTOM_OBJ, 4}};
+  size_t currentOrderIndex = 0;
+  bool seenNonPersistentBuffer = false;
+  for (const auto& inputSpec : inputSpecs) {
+    if (inputSpec.tag() == torch::_export::InputSpec::Tag::USER_INPUT ||
+        inputSpec.tag() == torch::_export::InputSpec::Tag::CONSTANT_INPUT) {
+      continue;
+    }
+    auto it = tagOrderMap.find(inputSpec.tag());
+    if (it == tagOrderMap.end()) {
+      TORCH_CHECK(false, "Unknown InputSpec tag encountered.");
+    }
+    size_t tagIndex = it->second;
+
+    if (tagIndex < currentOrderIndex) {
+      auto [argName, tagName] = getSpecDetails(inputSpec);
+      TORCH_CHECK(
+          false,
+          fmt::format(
+              "Input arg {} with InputSpec {} is out of order!",
+              argName,
+              tagName));
+    }
+    currentOrderIndex = tagIndex;
+    // Additional check for buffers
+    if (inputSpec.tag() == torch::_export::InputSpec::Tag::BUFFER) {
+      if (!inputSpec.get_buffer().get_persistent()) {
+        seenNonPersistentBuffer = true;
+      } else if (
+          inputSpec.get_buffer().get_persistent() && seenNonPersistentBuffer) {
+        TORCH_CHECK(
+            false,
+            "Persistent buffers must come before non-persistent buffers.");
+      }
+    }
+  }
+}
+
+void checkInputNames(
+    const std::unordered_set<std::string>& sigNames,
+    const std::unordered_set<std::string>& graphNames) {
+  if (sigNames == graphNames) {
+    return;
+  }
+
+  std::string errorMsg =
+      "Error: Value name difference detected between graph signature and graph nodes:\n";
+  errorMsg += "Signature value names:\n";
+  errorMsg += fmt::format("[{}]\n", fmt::join(sigNames, ", "));
+  errorMsg += "Graph node names:\n";
+  errorMsg += fmt::format("[{}]", fmt::join(graphNames, ", "));
```

Comment:
nit: why not use fmt::format once to construct the whole message? iteratively concatenating is inefficient